src/mrsimulator/simulator/__init__.py

```diff
@@ -320,6 +322,48 @@ def export_methods(self, filename: str):
                 mth, outfile, ensure_ascii=False, sort_keys=False, allow_nan=False
             )
 
+    def optimize(self) -> None:
+        """Pre-computes transition pathways and associated weights for each of the
+        :py:class:`~mrsimulator.method.Method` and
+        :py:class:`~mrsimulator.spin_system.SpinSystem` objects held by the simulator.
+        This increases the efficiency during least-squared minimization since pathways
+        are not re-computed during every iteration.
+
+        Example
+        -------
+
+        >>> sim = Simulator()
+        >>> # Add spin systems and methods
+        >>> sim.optimize()
+        >>> sim.run()
+        """
+        # NOTE: Could possibly make this into a list comprehension
+        for mth in self.methods:
+            pathways = []
+            weights = []
+            for sys in self.spin_systems:
+                p, w = mth._get_transition_pathway_and_weights_np(sys)
+                pathways.append(p)
+                weights.append(w)
+
+            self._precomputed_pathways.append(pathways)
+            self._precomputed_weights.append(weights)
```

Comment:
To avoid accidental release or forgot to release, it might be better to have a separate optimizer objects (a dict will do for now). The workflow would be:
```py
opt = sim.optimize()
sim.run(optimizer=opt)
```
Here, `opt` will be a dict object containing pre-computed pathways as weight.

This way the optimizer will have to be explicitly added as the argument. If the argument is missing, the sim.run() will calculate these internally.

```py
def sim(self, opt=None):
    if opt is None:
        opt = self.optimize()
```